### ISSUE TITLE
reshape the output of LSTM 2D Conv

### ISSUE THREAD

**Issue by HuangBo-Terraloupe** [keras-team/keras#9962]
First of all, the implementation 'kld 'of keras is not the real kld, it will get the negative value if the y_true is bigger than y_pred. But by definition, the kld can not be negative.
I got one implementation using tensorflow, 

def kl_divergence(labels, prediction, epsilon=1e-7):
   prediction /= (tf.reduce_sum(prediction, axis=1, keep_dims=True) + epsilon)
   labels /= (tf.reduce_sum(labels, axis=1, keep_dims=True) + epsilon)
   result = tf.reduce_mean(tf.reduce_sum(labels * tf.log((labels / (prediction + epsilon)) + epsilon), axis=1))
   return result

In order to use this function, y_true and y_pred need to be a 2D vector, shape=(batch_size, vector_length).

I have a LSTM model:

def lstm_model_2D(batch_size, time_step, input_size, output_size):

    model = Sequential()
    model.add(ConvLSTM2D(8, (3, 3), strides=(1, 1), padding='same', return_sequences=True, dropout=0.5,
                         recurrent_dropout=0.0, batch_size=batch_size, input_shape=(time_step, ) + input_size))

    model.add(ConvLSTM2D(4, (3, 3), strides=(1, 1), padding='same', return_sequences=True, dropout=0.5,
                         recurrent_dropout=0.0, batch_size=batch_size))

    model.add(ConvLSTM2D(1, (3, 3), strides=(1, 1), padding='same', return_sequences=True, dropout=0.5,
                         recurrent_dropout=0.0, batch_size=batch_size))
    model.add(TimeDistributed(Flatten()))
    return model

![screenshot from 2018-04-17 18-41-40](https://user-images.githubusercontent.com/26878250/38884164-02ed0f24-426f-11e8-8dc9-6532931145e0.png)

My question is:
how to reshape the output from (40, 32, 880) to (40x32, 880), so that I can use the 'kld' function above.
I tried keras.layer.Reshape(), it can only reshape the time_step(32), and the feature(880), it can not reshape the batch_size(40).

**Comment by HuangBo-Terraloupe** [comment]
Hello everyone, 
I tried my best to go into the details, I found the problem is: I need do softmax 2d and also calculate the 'kld' 2d. 
1, I trained a static model using softmax and 'kld', it works good, because the input to softmax and kld are 2d. The shape is (batch_size, 880).
Because I am doing a probability regression, so I need normalize every feature between value 0-1.
2, Because I am doing video frame regression, so I want to combine time information to my model. So I go into LSTM, the problem is, my output changes to 3d: (batch_size, time_step, length_feature).

The keras softmax and kld work different with the 3d input, I need somehow modify the code to make it works in 2d.

(1), I tried modify the LSTM model using keras.layers.Reshape, it can not operate on batch_size.
(2), I tried do it in loss function:
    def kullback_leibler_divergence(y_true, y_pred):
    '''reshape the 3d feature into 2d, so the calculate of softmax and kld works fine.
    Args:
        y_true: the groundtruth, shape = (batch_size, time_step, length_feature)
        y_pred: the prediction, shape = (batch_size, time_step, length_feature)

    Returns:

    '''

    tensor_shape = K.shape(y_true) # shape = (40, 32, 880)
    y_true = K.reshape(y_true, shape=[-1, tensor_shape[-1]]) # shape = (40*32, 880)
    y_pred = K.reshape(y_pred, shape=[-1, tensor_shape[-1]]) # shape = (40*32, 880)

    # apply 2d softmax to prediction, normalise the feature to a probability distribution
    y_pred = K.softmax(y_pred)

    y_true = K.clip(y_true, K.epsilon(), 1)
    y_pred = K.clip(y_pred, K.epsilon(), 1)
    return K.mean(K.sum(y_true * K.log(y_true / y_pred), axis=-1))
It works! 
I am not sure if this way is fine, I want to get some comments from the expert, thanks in advance!

**Comment by HuangBo-Terraloupe** [comment]
![screenshot from 2018-04-19 19-13-49](https://user-images.githubusercontent.com/26878250/39008182-403181e0-4408-11e8-9df8-8e80a19ccb30.png)

I solve the reshape problem, but seems keras does not support it for training.
The reason why I need reshape the output to a 2D tensor is, if I use 3D tensor to calculate the kld loss, I got wrong numbers and the evaluation performance is really bad. Can someone point out  what should I do?
Now I tried 2 ways, modify the loss function and modify the model structure, both does not work.
